### PR TITLE
Fix warnings

### PR DESCRIPTION
**Pull request by sodevel**
Fixes various warnings reported by MSVC 2015 except wxWidgets deprecation warnings which are handled by #417. Bigger rework of the MD5 class to prevent implicit type conversions.

### CODE REVIEW

File: plugins/additional/additional.cpp [hunk 1]
```diff
@@ -1320,12 +1320,12 @@ class SearchCtrlComponent : public ComponentBase
 
 		if ( !obj->IsNull( _("search_button") ) )
 		{
-			sc->ShowSearchButton( obj->GetPropertyAsInteger( _("search_button") ) );
+			sc->ShowSearchButton( !!obj->GetPropertyAsInteger( _("search_button") ) );
```
Comment (jhasse) [inline plugins/additional/additional.cpp]: I think `!= 0` would be clearer.

File: plugins/additional/additional.cpp [hunk 2]
```diff
@@ -1806,24 +1806,24 @@ void ComponentEvtHandler::OnMarginClick( wxStyledTextEvent& event )
 class DataViewModel : public wxDataViewModel
 {
 public:
-	unsigned int GetChildren(const wxDataViewItem& item, wxDataViewItemArray& children) const {
+	unsigned int GetChildren(const wxDataViewItem& WXUNUSED(item), wxDataViewItemArray& WXUNUSED(children)) const {
```
Comment (jhasse) [inline plugins/additional/additional.cpp]: Are there any advantages of `WXUNUSED` over just commenting out the name with `/* */`?

File: src/md5/md5.cc [hunk 3]
```diff
@@ -80,7 +80,7 @@ void MD5::update ( const uint1 *input, uint4 input_length) {
   }
 
   // Compute number of bytes mod 64
-  buffer_index = (unsigned int)((count[0] >> 3) & 0x3F);
+  buffer_index = (uint4)((count[0] >> 3) & 0x3F);
```
Comment (jhasse) [inline src/md5/md5.cc]: `uint32_t` would be even better. Why change the type at all though?

File: src/md5/md5.cc [hunk 4]
```diff
@@ -119,11 +119,11 @@ void MD5::update ( const uint1 *input, uint4 input_length) {
 
 void MD5::update(FILE *file){
 
-  unsigned char buffer[1024];
-  int len;
+  unsigned char fileBuffer[1024];
+  size_t len;
```
Comment (jhasse) [inline src/md5/md5.cc]: Why not `uint4`/`uint32_t`? Then you wouldn't need the `static_cast`.

File: src/md5/md5.cc [hunk 5]
```diff
@@ -119,11 +119,11 @@ void MD5::update ( const uint1 *input, uint4 input_length) {
 
 void MD5::update(FILE *file){
 
-  unsigned char buffer[1024];
-  int len;
+  unsigned char fileBuffer[1024];
```
Comment (jhasse) [inline src/md5/md5.cc]: Any reason for the rename?

File: src/md5/md5.cc [hunk 6]
```diff
@@ -219,6 +198,15 @@ void MD5::finalize (){
 
 
 
+MD5::MD5(const unsigned char* input, uint4 input_length) {
+	init();  // must be called be all constructors
```
Comment (jhasse) [inline src/md5/md5.cc]: Typo: by

File: src/rad/wxfbmanager.cpp [hunk 7]
```diff
@@ -155,16 +155,16 @@ wxObject* wxFBManager::GetWxObject( PObjectBase obj )
 
 void wxFBManager::ModifyProperty( wxObject* wxobject, wxString property, wxString value, bool allowUndo )
 {
-	CHECK_VISUAL_EDITOR()
+	CHECK_VISUAL_EDITOR(void())
```
Comment (jhasse) [inline src/rad/wxfbmanager.cpp]: Why is this needed?

File: plugins/additional/additional.cpp (hunk 1 above)
Comment (sodevel) [reply plugins/additional/additional.cpp]: You can debate if the double bang is clearer, but it is shorter and more flexible to normalize X to a bool. It can also be used in cases where X can't be compared to an int, e.g. the safe bool idiom or a ``wxStreamBase`` that doesn't convert to a bool at all and only has ``operator!``.

File: plugins/additional/additional.cpp (hunk 2 above)
Comment (sodevel) [reply plugins/additional/additional.cpp]: Portability. I don't know how much this is an issue today, but i remember having read that some compilers don't support this and require other means to ignore unused parameters. Thats why WXUNUSED, boost::ignore_unused() and the like exist, so i used the wxWidgets style here.

File: src/md5/md5.cc (hunk 3 above)
Comment (sodevel) [reply src/md5/md5.cc]: The type usage in this class was a huge mess, i tried the minimum change approach to fix it. ``buffer_index`` is a ``uint4`` so i cast the right side to the left side. Most probably this cast isn't necessary at all because ``count`` is a ``uint4`` as well and the integer literals are so small that they won't introduce a type cast at all, but the cast was there so i kept it.

File: src/md5/md5.cc (hunk 4 above)
Comment (sodevel) [reply src/md5/md5.cc]: Because ``fread`` returns a ``size_t`` which may be one of these but also can be ``uint64_t`` or whatever ``typedef /*implementation-defined*/ size_t;`` is in the used implementation. So in these locations i used the types native to the functions that write them and cast them to the types the functions need that read them.

File: src/md5/md5.cc (hunk 5 above)
Comment (sodevel) [reply src/md5/md5.cc]: To get rid of the shadow warnings caused by the class member with the same name. This is the reason for the various other renames as well.

File: src/md5/md5.cc (hunk 6 above)
Comment (sodevel) [reply src/md5/md5.cc]: I shouldn't have copy-pasted the file-constructor but the stream-constructor, there its correct...

File: src/rad/wxfbmanager.cpp (hunk 7 above)
Comment (sodevel) [reply src/rad/wxfbmanager.cpp]: This macro requires a parameter and MSVC spills out a warning if you don't supply one. There a tricks to produce macros that accept a variable amount of parameters or you can make two macros, one for methods that return something and one for methods that return void (wxWidgets uses this for its assert macros), but this was the least-amount-of-work solution.

Comment (jhasse) [reply src/rad/wxfbmanager.cpp]: Can you disable the warning for MSVC instead?

File: plugins/additional/additional.cpp (hunk 2 above)
Comment (jhasse) [reply plugins/additional/additional.cpp]: Okay, I think that isn't an issue today anymore.

File: plugins/additional/additional.cpp (hunk 1 above)
Comment (jhasse) [reply plugins/additional/additional.cpp]: In this case it actually *is* an integer though.

File: src/rad/wxfbmanager.cpp (hunk 7 above)
Comment (sodevel) [reply src/rad/wxfbmanager.cpp]: Interesting, according to the standard it is even an error to supply less arguments than required, but empty arguments are ok. For some reason [gcc](https://gcc.gnu.org/onlinedocs/cpp/Macro-Arguments.html) interprets empty parentheses as one empty argument but MSVC not, that one however doesn't produce an error if you don't supply enough.

Regarding that these macros are only used local in that file i guess i can disable the warning instead.

File: plugins/additional/additional.cpp (hunk 2 above)
Comment (sodevel) [reply plugins/additional/additional.cpp]: It is/was wxWidgets style to suppress these warnings, i would keep them so its easier in the future to replace them with [[maybe_unused]]

File: plugins/additional/additional.cpp (hunk 1 above)
Comment (sodevel) [reply plugins/additional/additional.cpp]: Still it is shorter and has its semantic meaning of "normalize-to-bool-operator". I am used to use it in my codebase, so what shall i do?

**Comment by sodevel** [comment]
I pushed a replacement for df72141 but kept it. I can rebase this PR to current master to fix that trivial merge conflict and remove that commit and do other cleanup like maybe splitting the MD5 refactoring which is a little more than just a warning fix but for now i keep it as is to prevent issues with github.

File: plugins/additional/additional.cpp (hunk 2 above)
Comment (jhasse) [reply plugins/additional/additional.cpp]: `[[maybe_unused]]` is even longer and "maybe" just seems like "disable the warning" to me. But it's a good warning because it can catch some problems (e.g. mistyping) or unnecessary parameters.

In this case, I think it's actually best to just remove the parameter name, as "item" doesn't add any extra information anyway.

I've fixed the warning with `/* */` or by removing the name in 0985941f5e4dac3c1a46ae029cbb66cd257d1b97.

File: src/rad/wxfbmanager.cpp [hunk 8]
```diff
@@ -26,6 +26,12 @@
 #include "model/objectbase.h"
 #include "rad/appdata.h"
 
+// Sometimes these macros get called with an empty argument which is not an error
+// but MSVC still produces a warning about too few arguments, disable it
+#ifdef _MSC_VER
+#pragma warning(disable:4003)
```
Comment (jhasse) [inline src/rad/wxfbmanager.cpp]: You can also disable warnings in MSVC's project files, which I would prefer, as I don't like to add MSVC specific code in cpp files.

(I actually don't like to support MSVC at all, as it increases the maintenance burden and is proprietary)

**Comment by jhasse** [comment]
I've cherry-picked a few commits from this PR to the master branch (messed it up actually, so CI failed. Sorry about that!).

It might be a little bit harder for you to rebase this now. Sorry again ...

File: src/rad/wxfbmanager.cpp (hunk 8 above)
Comment (sodevel) [reply src/rad/wxfbmanager.cpp]: Since MSVC is like the "native" compiler on windows i am quite interested that it is supported. Plus compiling wxWidgets on my machine with MSVC running on one core is still faster than with gcc running on all four.

**Comment by sodevel** [comment]
Rebased on current master

File: src/md5/md5.cc [hunk 9]
```diff
@@ -215,31 +199,32 @@ void MD5::finalize (){
...
 
 
 
-MD5::MD5(FILE *file){
-
-	init(); // must be called by all constructors
-  update(file);
-  finalize ();
+MD5::MD5(const unsigned char* input, uint32_t input_length) {
+	init();  // must be called be all constructors
```
Comment (jhasse) [inline src/md5/md5.cc]: Introduces the typo again.

File: src/md5/md5.cc [hunk 10]
```diff
@@ -215,31 +199,32 @@ void MD5::finalize (){
...
 
 
-MD5::MD5(istream& stream){
+MD5::MD5(FILE *file){
 
 	init(); // must be called by all constructors
-  update (stream);
-  finalize();
+  update(file);
+  finalize ();
```
Comment (jhasse) [inline src/md5/md5.cc]: Please remove the space.

File: src/md5/md5.cc [hunk 11]
```diff
@@ -250,11 +235,9 @@ MD5::MD5(ifstream& stream){
 
 unsigned char *MD5::raw_digest(){
 
-	uint8_t* s = new uint8_t[16];
+  unsigned char *s = new unsigned char[16];
 
   if (!finalized){
-    cerr << "MD5::raw_digest:  Can't get digest if you haven't "<<
-      "finalized the digest!" <<endl;
```
Comment (jhasse) [inline src/md5/md5.cc]: Aren't those error messages a good thing? Now the caller doesn't have an idea why it failed.

File: plugins/additional/additional.cpp (hunk 1 above)
Comment (jhasse) [reply plugins/additional/additional.cpp]: If it isn't okay to implicitly cast an int to bool, I don't think that passing an int to `!` should be allowed either.

> safe bool idiom

If I understand it correctly this is obsolete now that C++11 supports explicit conversion operators.

File: build/premake/wxformbuilder.lua [hunk 12]
```diff
@@ -72,6 +72,7 @@ project "wxFormBuilder"
 
     configuration "vs*"
         defines         {"_CRT_SECURE_NO_DEPRECATE", "_CRT_SECURE_NO_WARNINGS"}
+        buildoptions    {"/wd4003"}
```
Comment (jhasse) [inline build/premake/wxformbuilder.lua]: Perfect, thanks :)

File: src/md5/md5.cc (hunk 9 above)
Comment (sodevel) [reply src/md5/md5.cc]: I had so much fun with interactive rebasing and editing, squashing, splitting commits, i guess some fun slipped through the cage, will fix this later

File: src/md5/md5.cc (hunk 11 above)
Comment (sodevel) [reply src/md5/md5.cc]: This is sort of a library and it shouldn't produce unexpected console output. Plus this doesn't work with a windows GUI application anyway. Plus this doesn't help you to catch this error programmatically.

On second thought, this isn't even a real error condition but a programmer error that does only deserve an assertion, like some code below already does. It's a much more serious error to actually ignore the read errors from the FILE/stream.

It's just amazing how many issues are in this tiny piece of code, i just wanted to fix some warnings...

File: plugins/additional/additional.cpp (hunk 1 above)
Comment (sodevel) [reply plugins/additional/additional.cpp]: This doesn't stop [MSVC](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4800) to complain about

Comment (sodevel) [reply plugins/additional/additional.cpp]: I see you don't like my double bang operator, i could go the /wd4800 route if you prefer

Comment (jhasse) [reply plugins/additional/additional.cpp]: > This doesn't stop MSVC to complain about

What do you mean by "this"?

> I see you don't like my double bang operator, i could go the /wd4800 route if you prefer

I still think `!= 0` would be a good idea, it will also silence MSVC's warning. I will push something tomorrow.

File: src/md5/md5.cc (hunk 11 above)
Comment (jhasse) [reply src/md5/md5.cc]: I see. What do you think about throwing a `std::runtime_error`?

File: plugins/additional/additional.cpp (hunk 1 above)
Comment (sodevel) [reply plugins/additional/additional.cpp]: > What do you mean by "this"?

Explicit conversion

File: src/md5/md5.cc (hunk 11 above)
Comment (sodevel) [reply src/md5/md5.cc]: The exception for what? Getting called in the wrong finalization state? Or the read errors? I wouldn't use an exception here for both cases, this whole "library" is exception free and looks so much c-like and the calling side is not prepared for this, its not worth the trouble. Assert for the finalization and if necessary use a return code for read errors.

Comment (jhasse) [reply src/md5/md5.cc]: The calling side isn't prepared for an assert either though. And what "trouble" do you mean?

Comment (sodevel) [reply src/md5/md5.cc]: Maybe there is a misunderstanding what an assert is, so lets clear this up: In release mode an assert does nothing, in debug mode, if it triggers, it displays the message and then terminates the program, so there is no calling side anymore ;).

wxWidgets has these nice "hybrid-asserts", they are also active in release mode and silently perform the trigger-action there, in debug mode they pop up the message and let you choose what to do. These are a "nicer" version of a "true" assert, they allow the program to continue with some meaningful value in cause of the assert.

The "trouble" with exceptions here is that they are not so common like in java and here there is also no such thing like checked exceptions. So you have to write some documentation like "This does throw this and that exception in case of error". The TiCPP library does this, wxFB itself uses internally its own exception class, but this class is so small, does nasty c-style things like returning something you need to delete, basically the only c++-element is the istream (ok, and ostream inserter), an error code would fit here much better in my opinion.

With some small modifications you don't even need to signal read errors at all, you could let the caller check the error state of the stream to detect an error (currently it always will be in error state because of the way it is read from, but this can be made better). For the FILE version you need to stop to delete it in the method and it would work there as well. That would give a nice interface that is identical to the memory block version.

**Comment by jhasse** [comment]
While some changes could be squashed into fewer commits, I think that's nitpicking at this point. Thank you very much for the changes, highly appreciated!
